### PR TITLE
Some rustups

### PR DESCRIPTION
**Pull request by JohnTitor**
changelog: none

### CODE REVIEW

File: clippy_lints/src/utils/author.rs [hunk 1]
```diff
@@ -617,9 +617,13 @@ impl<'tcx> Visitor<'tcx> for PrintVisitor {
                     start_pat, end_pat, end_kind, current
                 );
                 self.current = start_pat;
-                self.visit_expr(start);
+                if let Some(expr) = start {
+                    self.visit_expr(expr);
```
Comment (Centril) [inline clippy_lints/src/utils/author.rs]: btw, there's a nice `syntax::walk_list!(self, visit_expr, start)` pattern for this sorta thing.

File: clippy_lints/src/matches.rs [hunk 2]
```diff
@@ -713,16 +713,18 @@ fn all_ranges<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, arms: &'tcx [Arm<'_>]) -> Ve
...
-                    let rhs = constant(cx, cx.tables, rhs)?.0;
-                    let rhs = match *range_end {
-                        RangeEnd::Included => Bound::Included(rhs),
-                        RangeEnd::Excluded => Bound::Excluded(rhs),
-                    };
-                    return Some(SpannedRange {
-                        span: pat.span,
-                        node: (lhs, rhs),
-                    });
+                    if let (Some(l), Some(r)) = (lhs, rhs) {
```
Comment (Centril) [inline clippy_lints/src/matches.rs]: (Leaving this comment as a motivational example for https://github.com/rust-lang/rust/issues/65490.)

Comment (Centril) [reply clippy_lints/src/matches.rs]: Also, is there a reason not to support `None` at each end here?

File: clippy_lints/src/utils/hir_utils.rs [hunk 3]
```diff
@@ -195,7 +195,10 @@ impl<'a, 'tcx> SpanlessEq<'a, 'tcx> {
                 ls == rs && over(l, r, |l, r| self.eq_pat(l, r))
             },
             (&PatKind::Range(ref ls, ref le, ref li), &PatKind::Range(ref rs, ref re, ref ri)) => {
-                self.eq_expr(ls, rs) && self.eq_expr(le, re) && (*li == *ri)
+                if let (Some(ls), Some(rs), Some(le), Some(re)) = (ls, rs, le, re) {
```
Comment (Centril) [inline clippy_lints/src/utils/hir_utils.rs]: Not sure why you're not supporting half-open ranges here.

File: clippy_lints/src/matches.rs (hunk 2 above)
Comment (JohnTitor) [reply clippy_lints/src/matches.rs]: I'm not sure how to pass the `Expr`s to `constant()` on half patterns, any thoughts?

Comment (Centril) [reply clippy_lints/src/matches.rs]: This is what the compiler does: https://github.com/rust-lang/rust/pull/67258/files#diff-e826542d6f24137d5b0b5ae333230529R493-R511

File: clippy_lints/src/utils/hir_utils.rs (hunk 3 above)
Comment (JohnTitor) [reply clippy_lints/src/utils/hir_utils.rs]: We should also support some patterns like `(Some(ls), None, None, Some(re))` here?

Comment (Centril) [reply clippy_lints/src/utils/hir_utils.rs]: Well, surely e.g., `1..` and `1..` should compare equal, so you need something like:
```rust
both(ls, rs, |a, b| self.eq_expr(a, b))
&& both(le, re, |a, b| self.eq_expr(a, b))
&& (*li == *ri)
```

**Comment by JohnTitor** [comment]
@Centril Added commit 10cf141ebb9c9856858d5f4cc36281b2b43d37b3, could you review it?

**Comment by JohnTitor** [comment]
I noticed we need more rustup, I'm preparing now (it needs large import pass changes...)

File: clippy_lints/src/matches.rs [hunk 4]
```diff
@@ -705,25 +705,52 @@ fn check_wild_in_or_pats(cx: &LateContext<'_, '_>, arms: &[Arm<'_>]) {
...
+                                RangeEnd::Included => Bound::Included(rhs),
+                                RangeEnd::Excluded => Bound::Excluded(rhs),
+                            };
+                            return Some(SpannedRange {
+                                span: pat.span,
+                                node: (lhs, rhs),
+                            });
+                        },
+                        (None, Some(rhs)) => {
+                            let lhs = miri_to_const(ty.numeric_min_val(cx.tcx)?)?;
```
Comment (Centril) [inline clippy_lints/src/matches.rs]: A simpler approach could be to use `{l,r}hs.unwrap_or_else(|| ...)` and simply ignore the fact that `(None, None)` is a possibility.

Comment (JohnTitor) [reply clippy_lints/src/matches.rs]: Uhm, we can use `map_or` but it requires some `unwrap`s, I think.
like this:

```rust
let lhs = lhs.map_or(
    miri_to_const(ty.numeric_min_val(cx.tcx).unwrap()).unwrap(),
    |l| constant(cx, cx.tables, l).unwrap().0
);
let rhs = rhs.map_or(
    miri_to_const(ty.numeric_min_val(cx.tcx).unwrap()).unwrap(),
    |r| constant(cx, cx.tables, r).unwrap().0
);
```

Is there more appropriate approach?

**Comment by JohnTitor** [comment]
Uhm, fmt doesn't know half-open ranges?

File: clippy_lints/src/matches.rs (hunk 4 above)
Comment (Centril) [reply clippy_lints/src/matches.rs]: Well you can use `match` on `lhs` or `rhs` individually, and it will still be an improvement.

Comment (JohnTitor) [reply clippy_lints/src/matches.rs]: You suggest to collapse `(None, Some(rhs))` and `(Some(lhs), None)`, right?

Comment (Centril) [reply clippy_lints/src/matches.rs]: My suggestion is to do:
```rust
let lhs = match lhs {
    Some(expr) => constant(cx, cx.tables, lhs)?.0,
    None => miri_to_const(ty.numeric_min_val(cx.tcx)?)?,
};
```
and similarly with `rhs`, but with `numeric_max_val`.

Comment (JohnTitor) [reply clippy_lints/src/matches.rs]: Ah thanks, I've just got your intention, reduced lines via 3bdf40407606b82becacdb1c0f77e73c2f6c9871.

**Comment by flip1995** [comment]
Can you try to add a `#[rustfmt::skip]` to the test (with a FIXME)?

**Comment by JohnTitor** [comment]
> Can you try to add a `#[rustfmt::skip]` to the test (with a FIXME)?

Uhm, rustfmt still says it's unsupported on my local, should I push here?

**Comment by flip1995** [comment]
Hm, then I think the only thing we can do is to comment out the test and open an issue in rustfmt.

**Comment by flip1995** [comment]
@bors delegate+ p=10

r=me once Travis passes. I have to go now, and won't have access to a PC until tomorrow evening.

**Comment by JohnTitor** [comment]
Ugh, just needs more rustup, okay, I'll fix soon...

**Comment by JohnTitor** [comment]
Okay, latest rustup is added, I'll r=flip1995 once CI is green.

**Comment by JohnTitor** [comment]
Removed references, let's try again.
@bors r=flip1995

**Pull request merged**